share-coordinator/src/test/java/org/apache/kafka/coordinator/share/StateBatchUtilTest.java

```diff
@@ -0,0 +1,472 @@
+
+        BatchTestHolder(String testName,
+                        List<PersisterStateBatch> curList,
+                        List<PersisterStateBatch> newList,
+                        List<PersisterStateBatch> expectedResult,
+                        long startOffset) {
+            this(testName, curList, newList, expectedResult, startOffset, true);
+        }
+
+        BatchTestHolder(String testName,
+                        List<PersisterStateBatch> curList,
+                        List<PersisterStateBatch> newList,
+                        List<PersisterStateBatch> expectedResult,
+                        long startOffset,
+                        boolean shouldRun) {
+            this.testName = testName;
+            this.curList = curList;
+            this.newList = newList;
+            this.expectedResult = expectedResult;
+            this.startOffset = startOffset;
+            this.shouldRun = shouldRun;
+        }
+
+        @Override
+        public String toString() {
+            return this.testName;
+        }
+    }
+
+    @SuppressWarnings({"MethodLength"})
+    private static Stream<BatchTestHolder> generator() {
+        return Stream.of(
+            new BatchTestHolder(
+                "Current batches with start offset midway are pruned.",
+                Collections.singletonList(
+                    new PersisterStateBatch(100, 130, (byte) 0, (short) 1)
+                ),
+                Collections.emptyList(),
+                Collections.singletonList(
+                    new PersisterStateBatch(120, 130, (byte) 0, (short) 1)
+                ),
+                120
+            ),
+
+            new BatchTestHolder(
+                "New batches with start offset midway are pruned.",
+                Collections.emptyList(),
+                Collections.singletonList(
+                    new PersisterStateBatch(100, 130, (byte) 0, (short) 1)
+                ),
+                Collections.singletonList(
+                    new PersisterStateBatch(120, 130, (byte) 0, (short) 1)
+                ),
+                120
+            ),
+
+            new BatchTestHolder(
+                "Both current and new batches empty.",
+                Collections.emptyList(),
+                Collections.emptyList(),
+                Collections.emptyList(),
+                120
+            ),
+
+            // same state
+            new BatchTestHolder(
+                "Same state. Last and candidate have same first and last offset.",
+                Collections.singletonList(
+                    new PersisterStateBatch(100, 110, (byte) 0, (short) 1)
+                ),
+                Collections.singletonList(
+                    new PersisterStateBatch(100, 110, (byte) 0, (short) 1)
+                ),
+                Collections.singletonList(
+                    new PersisterStateBatch(100, 110, (byte) 0, (short) 1)
+                ),
+                -1
+            ),
```

Comment:
Since you have so many of these, I would suggest a factory that lets you pass in ints. 
E.g., 
```
new BatchTestHolder(
  "Same state. Last and candidate have same first and last offset.",
  BatchTestHolder.singleBatch(100, 110, 0, 1),
  BatchTestHolder.singleBatch(100, 110, 0, 1),
  ...
);

```